[Navigation] Fix cross-window replace navigations

TingPing opened the pull request:
#### e2da0c499986dc9703d22df4b7b1a15020087306
<pre>
[Navigation] Fix cross-window replace navigations
<a href="https://bugs.webkit.org/show_bug.cgi?id=282398">https://bugs.webkit.org/show_bug.cgi?id=282398</a>

Reviewed by Alex Christensen.

* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-origin-traversal-redirect-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-replace-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/navigation-timing/nav2_test_redirect_xserver-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/navigation-timing/test_timing_xserver_redirect-expected.txt:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::initializeForNewWindow):

Canonical link: <a href="https://commits.webkit.org/286115@main">https://commits.webkit.org/286115@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c2693d81c81fe7d76bdbfd130f3b11c28de2838

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74822 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54252 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27639 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79277 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26060 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76939 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63388 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2037 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58787 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17053 "20 failures") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77889 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48928 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64295 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39179 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46206 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21792 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24393 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67374 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22137 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80736 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2140 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1288 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67027 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2288 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64313 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66320 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16482 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10267 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8415 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2105 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4892 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2133 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3054 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2140 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -135,9 +135,20 @@ void Navigation::initializeForNewWindow(std::optional<NavigationNavigationType>
+                if (navigationType == NavigationNavigationType::Replace)
+                    m_entries[*previousNavigation->m_currentEntryIndex] = NavigationHistoryEntry::create(protectedScriptExecutionContext().get(), *currentItem);
+
+                m_currentEntryIndex = getEntryIndexOfHistoryItem(m_entries, *currentItem);
```
Agreed that it is better to not rely on previous m_currentEntryIndex but rather compute the latest value.

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -135,9 +135,20 @@ void Navigation::initializeForNewWindow(std::optional<NavigationNavigationType>
+
+                m_currentEntryIndex = getEntryIndexOfHistoryItem(m_entries, *currentItem);
+
+                if (navigationType)
```
I think I asked before, but when is this null? And do we have to test?

rwlbuis replied:
Note, possibly is 204/205.

TingPing replied:
Removed for now. Will see if any test hits this later.

TingPing replied:
A handful of tests do fail because of this, will readd and see if it being unset in those cases makes sense.

TingPing replied:
The failures were all forms/POST requests:

```
fast/history/form-submit-in-frame-via-onclick.html
fast/history/form-submit-in-frame.html
http/tests/navigation/post-frames.html
http/tests/navigation/postredirect-frames.html
```

I re-added this with a comment.

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -135,9 +135,20 @@ void Navigation::initializeForNewWindow(std::optional<NavigationNavigationType>
+                m_currentEntryIndex = getEntryIndexOfHistoryItem(m_entries, *currentItem);
+
+                if (navigationType)
+                    m_activation = NavigationActivation::create(*navigationType, *currentEntry(), WTFMove(previousEntry));
```
Is updateForActivation not being called because it does too much work or incorrect work for this code path?

TingPing replied:
All of the work in `updateForActivation` is to try and figure out the previousEntry.

In this case we just have it. So its literally one call to make the activation.

Passing an Entry directly instead of making a new one from an Item does help fix the tests here.

rwlbuis reviewed (commented):
LGTM!

rwlbuis reviewed (commented):
LGTM again, thanks for checking the null navigationType issue.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -135,9 +135,20 @@ void Navigation::initializeForNewWindow(std::optional<NavigationNavigationType>
+                Ref previousEntry = m_entries[*previousNavigation->m_currentEntryIndex];
+
+                if (navigationType == NavigationNavigationType::Replace)
+                    m_entries[*previousNavigation->m_currentEntryIndex] = NavigationHistoryEntry::create(protectedScriptExecutionContext().get(), *currentItem);
```
This seems like it desperately needs a bounds check.  It is not clear to me that this isn't introducing a possible out-of-bounds write.  And a read on line 142.

TingPing committed:
[Navigation] Fix cross-window replace navigations
https://bugs.webkit.org/show_bug.cgi?id=282398

Reviewed by Alex Christensen.

* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-origin-traversal-redirect-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-replace-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/navigation-timing/nav2_test_redirect_xserver-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/navigation-timing/test_timing_xserver_redirect-expected.txt:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::initializeForNewWindow):

Canonical link: https://commits.webkit.org/286115@main
